src/api/controllers/db-controller.ts

```diff
@@ -828,43 +929,183 @@ export async function getTxFromDataStore(
   };
 }
 
+function parseContractWithAssociatedTx(
+  contracts: DbSmartContract[],
+  dbTxs?: DbTx[],
+  dbMempoolTx?: DbMempoolTx[]
+): Transaction[] {
+  const transactions: Transaction[] = [];
+  if (dbTxs) {
+    contracts.forEach(contract => {
+      const dbTx = dbTxs.find(tx => tx.contract_call_contract_id === contract.contract_id);
+      if (dbTx) {
+        const transaction = parseContractCallMetadata({ found: true, result: contract }, dbTx);
+        if (transaction) {
+          transactions.push(transaction as Transaction);
+        }
+      }
+    });
+  } else if (dbMempoolTx) {
+    contracts.forEach(contract => {
+      const dbMempool = dbMempoolTx.find(
+        tx => tx.contract_call_contract_id === contract.contract_id
+      );
+      if (dbMempool) {
+        const transaction = parseContractCallMetadata(
+          { found: true, result: contract },
+          undefined,
+          dbMempool
+        );
+        if (transaction) {
+          transactions.push(transaction as Transaction);
+        }
+      }
+    });
+  }
+  return transactions;
+}
+
+async function getContractCallForTxsList(
+  db: DataStore,
+  dbTxs?: DbTx[],
+  dbMempoolTxs?: DbMempoolTx[]
+): Promise<Transaction[] | undefined> {
+  // get contract call ids
+  const transactions = dbTxs ?? dbMempoolTxs ?? undefined;
+  if (transactions === undefined) {
+    return transactions;
+  }
+  const contractCallIds: string[] = [];
+  transactions.forEach((transaction: DbMempoolTx | DbTx) => {
+    if (transaction && transaction.contract_call_contract_id)
+      contractCallIds.push(transaction.contract_call_contract_id);
+  });
+  const contracts = await db.getSmartContractList(contractCallIds);
+  return parseContractWithAssociatedTx(contracts, dbTxs, dbMempoolTxs);
+}
+
 async function getContractCallMetadata(
   db: DataStore,
-  dbTx: DbTx | DbMempoolTx,
-  parsedTx: Transaction | MempoolTransaction
-): Promise<void> {
+  dbTx?: DbTx,
+  dbMempoolTx?: DbMempoolTx
+): Promise<ContractCallTransaction | MempoolContractCallTransaction | undefined> {
   // If tx type is contract-call then fetch additional contract ABI details for a richer response
+  const dbTransaction = db ?? dbMempoolTx ?? undefined;
```

Comment:
Same comment here -- unclear why there are 2 optional types of tx arrays for args, then only one of them is used.